hail/python/hail/ir/ir.py

```diff
@@ -1389,6 +1389,9 @@ def __init__(self, elements):
     def copy(self, *args):
         return MakeTuple(args)
 
+    def head_str(self):
+        return f'({" ".join([str(i) for i in range(len(self.elements))])})'
```

Comment:
hmmm.

I think yes, since we do want to be able to serialize/deserialize the IR at any   stage of lowering and optimization.

We don't need the full general form in Python yet, though, I think.

Reply:
however, I'm now convinced I need to add non-canonical type parsing logic.

Reply:
oh, I already did. But now we have another test :)